worlds/tunic/__init__.py

```diff
@@ -412,12 +441,18 @@ def fill_slot_data(self) -> Dict[str, Any]:
             "disable_local_spoiler": int(self.settings.disable_local_spoiler or self.multiworld.is_race),
         }
 
+        self.player_item_link_locations = {}
+        for group, item_links in self.item_link_locations.items():
```

Comment:
Unfortunately, `stage_generate_output` (sets `item_link_locations`) and `fill_slot_data` (reads `item_link_locations`) are run multithreaded, so there is no guarantee of their execution order. The latest non-threaded world method is `stage_post_fill` which is unfortunately still too early.

`stage_generate_output` is submitted to the ThreadPoolExecutor:
https://github.com/ArchipelagoMW/Archipelago/blob/daad3d0350cffabe356a80dd2f815645dd485ca4/Main.py#L240

`write_multidata` (which calls `fill_slot_data`) is submitted to the ThreadPoolExecutor:
https://github.com/ArchipelagoMW/Archipelago/blob/daad3d0350cffabe356a80dd2f815645dd485ca4/Main.py#L351

There are some worlds that wait for one function to be completed before the other can finish (OoT and Zillion), but this meant they had to be excluded from the `test.general.test_implemented.test_slot_data` test and if TUNIC were to wait for `stage_generate_output` to finish, it would likely have to be excluded as well. Zillion's is much simpler, so if you do want to go this route, I'd recommend looking at Zillion's code over OoT's.

---

Zillion's process:

Initialised in `__init__()` (a difference here is that Zillion uses `generate_output`, whereas the current code in this PR uses `stage_generate_output`, so the threading.Event would likely want to be set on the World class rather than per-instance)
https://github.com/ArchipelagoMW/Archipelago/blob/daad3d0350cffabe356a80dd2f815645dd485ca4/worlds/zillion/__init__.py#L130

Set in `generate_output`
https://github.com/ArchipelagoMW/Archipelago/blob/daad3d0350cffabe356a80dd2f815645dd485ca4/worlds/zillion/__init__.py#L373

Wait in `fill_slot_data`
https://github.com/ArchipelagoMW/Archipelago/blob/daad3d0350cffabe356a80dd2f815645dd485ca4/worlds/zillion/__init__.py#L396

Reply:
I think it's probably possible to have the first TUNIC world to call `fill_slot_data` run what's currently in `stage_generate_output` and have later TUNIC worlds that call `fill_slot_data` instead wait for `item_link_locations` to be ready.

Reply:
> I think it's probably possible to have the first TUNIC world to call `fill_slot_data` run what's currently in `stage_generate_output` and have later TUNIC worlds that call `fill_slot_data` instead wait for `item_link_locations` to be ready.

Yes, this seems to work and avoids the issue of not being able to run the test_slot_data test: https://github.com/Mysteryem/Archipelago-ahit/commit/c61ec62ce4cb1b00a2e52a46554886f29b57ba5b